README.md

R34 stamp is designed around Microchip’s ATSAMR34J18 SiP,  a 32-bit Arm Cortex-M0+ MCU (256Kb of Flash, 40Kb of RAM, 8kb of RWW flash) with an internal SX1276 LoRa transceiver. The SiP has a maximum power output of 20.00dBm, a frequency range of 862 to 1020MHz, and a 790nA sleep current.

This core is an arduino port for the Microchip ATSAMR34 Series microcontrollers - ATSAML21 with integrated SX1276 LoRa Tranceivers

## RIOT R34 Stamp setup

Board definition for Arduino IDE to install packages :

JSON URL : https://raw.githubusercontent.com/riotnetwork/samr-json/master/package_riot_index.json

* Open your Arduino IDE (V1.6.4 or newer.)
* In Arduino, Click File and then Preferences
* Add the JSON URL to the Additional Boards Manager text box. These are separated by a comma if you are using more than one added board link.
* Click Ok to close and save the preferences
* Then click Tools > Board > Board Manager
* You should see the Manager update on the first time opening after changing the preferences.
* Once updated you should find Riot boards listed in the Board Manager window.
* Find the install button next to this Board class and install.

* Restart Arduino IDE and look under Tools > Boards and you should see the addition of the Riot boards in your board list.

## RIOT R34 Stamp
This core is an arduino port for the Microchip ATSAMR34 and R35 Series microcontrollers - ATSAML21 with integrated LoRa Tranceivers

![R34 pinout](https://github.com/riotnetwork/samr-json/blob/master/Pinouts/PNG/R34-Stamp.png)

## License and credits

This core has been developed by RIOT network based on works by Arduino LLC in collaboration with Atmel as well as some influences from [Wiz-IO](https://github.com/Wiz-IO/platform-sam-lora) .

```
  Copyright (c) 2015 Arduino LLC.  All right reserved.

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
  See the GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
```
#ToDo's
integrate with platformIO 


CHANGELOG
SAMR CORE 1.0.10
updated to use arduino API in samd core 1.8.11
merged Commits on Apr 6, 2021 on https://github.com/arduino/ArduinoCore-samd/commits/master

SAMR CORE 1.0.9
SAMR34 ( and SAML21 ) port of arduino samd
based on SAMD CORE 1.6.21 2019.04.01
SAMD CORE 1.6.21 2019.04.01

* MKR boards: changed I2C to sercom2, SPI1 + Serial2 to sercom4
* Improved accuracy of delay() function. Thanks @BenF
* MKR 1500: Changed SARA module to be powered off on boot

SAMD CORE 1.6.20 2018.11.28

* Replaced boolean type with bool in examples. Thanks @per1234
* Added c++ linker command to allow to include libstdc++ when linking. Thanks @helmut64
* CPX driver fixes. Thanks @dhalbert
* I2S: Changed library to use 8 MHz oscillator source if 48MHz divider does not fit in 8 bits
* UART: Added frame error handling
* USB: Fixed memory leak on reconnects
* SDU: Added support for Arduino M0. Thanks @jandrassy
* Added arduinoOTA upload keys for Arduino M0. Thanks @jandrassy
* USB: Fixed USB Host failures and fixed memory overwrite in USBHost. Thanks @gdsports
* USB: Added method to return USB error code. Thanks @MarkFischer
* CDC: Clear line state on end()
* USB: Added USB device end() method
* Removed requirement that the DAC is on A0. Thanks @GabrielNotman
* Added alternate ports 44, 45 to make the SWCLK and SWDIO pins available on the Zero. Thanks @helmut64
* Added defines for MKR pin layout
* Fixed freeze in tone()
* Added MKR NB 1500 variant and bootloader
* Increased the default serial buffer size to 256

SAMD CORE 1.6.19 2018.07.11

* Fixed bootloader tools for .org boards
* M0: Updated pin definitions for D6, D7 and D13 to match Zero
* SPI: Fixed interrupt mask to block. Thanks @ggajoch
* Added MKR WiFi 1010 variant and bootloader
* Updated Windows Drivers to 1.4.0 and re-signed Adafruit_Circuit_Playground_Express.inf

SAMD CORE 1.6.18 2018.03.05

* Wire: Added support for general call (broadcast)
* SPI: Added SPI.notUsingInterrupt(...) API
* Fixed build errors if variant disabled USB
* Wire: TX and RX buffers are now 256 bytes (previously was 64 bytes)
* Added samd21e sercom compatibility. Thanks @mitchellpontague
* Fixed lock ups when outputting to UART during ISR
* Wire: correct I2C frequency calculations, and allow variant to overide default pull up resistor rise time
* USB: Fixed build errors when CDC_ENABLED was not set. Thanks @gdsports

SAMD CORE 1.6.17 2017.11.29

* More accurate approximation for multiplication factor of PLL to make 48Mhz. Thanks @joverbee
* Fix changing trigger mode (RISING/FALLING/...) in attachInterrupt(..) during runtime. Thanks @joverbee
* Improved ISR response time. Thanks @joverbee
* USB power current can now be overridden in the variant. Thanks @awatterott
* UART's now support optional RTS and CTS pins defined in the variant.
* Added support for MKR GSM 1400 and MKR WAN 1300

SAMD CORE 1.6.16 2017.08.23

* Is now possible to use PWM on D7 (though it can't be used with PWM on D12 at the same time). Thanks @Adminius
* PWMs now can perform real 16-bit resolution if analogWriteResolution(16) is set. Thanks @Adminius
* USB CDC: fixed issue of available() getting stuck when receiving ZLP's
* Serial (UART) tx is now buffered.
* Improved CRYSTALLESS stability. Thanks @tannewt
* Some fixes to Adafruit Circuit Playground Express variant. Thanks @ladyada
* Updated Stream and Print class
* Native USB now supports USB Serial Number
* Fixed pgm_read_ptr compatibility macro. Thanks @nkrkv

SAMD CORE 1.6.15 2017.04.27

* Added missing Windows driver for MKRFox1200
* Added SDU support for MKRFox1200

SAMD CORE 1.6.14 2017.04.04

* Fixed platform definition for openocd
* Added bootloader definitions for MKRFox1200
* Added lowpower function on USB subsystem

SAMD CORE 1.6.13 2017.03.31

* Added SDU (SD card Updater) library.
* Fixed 8MHz clock being generated incorrectly. Thanks @keestux

SAMD CORE 1.6.12 2017.02.07

* Fixed type conflict of utoa function. Thanks @kbumsik
* Fixed pinMode(pin, OUTPUT) not working after analogRead(pin)
* Reenabled support for Tian
* Fixed signature for Adafruit Circuit Playground Express board
* Added CMSIS headers for SAML21 chip series. Thanks @aethaniel

SAMD CORE 1.6.11 2016.12.23

* Temporarily disabled support for Arduino Tian until an upload issue is fixed

SAMD CORE 1.6.10 2016.12.23

* Added .org boards
* Fixed MKRZero `build.board` property
* linker: .data section is now aligned to 16 bytes address
* Added support for CRYSTALLESS configurations (thanks @ladyada)
* Added Circuits Playground board
* Added method to automatically retrieve sketch start via linker variable
  (see https://github.com/arduino/ArduinoCore-samd/pull/185)
* Fixed USBDevice.epBank1EnableTransferComplete method
* Adapted bootloader source code to CMSIS 4.5

SAMD CORE 1.6.9 2016.12.12

* MKR1000: Enabled PWM on pin 11, it is shared with pin A4
* SPI: Optimized SPI.transfer(...)
* Added linker flags for the ARM Math library
* Added new I2S library
* Bootloader: Check sketch vectors before waiting for double tap
* Upgraded CMSIS version to 4.5.0, and use new CMSIS-Atmel package
* Added Arduino MKRZero board.

SAMD CORE 1.6.8 2016.10.12

* Fixed regression on analogWrite. Thanks @bose-mdellisanti!
* Fixed regression on SerialUSB.available()
* C++ global constructors (if any) are now run after hardware initalization.

SAMD CORE 1.6.7 2016.09.28

* Fixed Serial.flush() blocking before any writes. Thanks @hangmoh
* Added snprintf_P to avr/pgmspace.h stub. Thanks @jantje
* Another small fix to String iterators. Thanks @Ivan-Perez @Chris--A
* Fixes SerialUSB.write(...) returning 0 instead of byte written
* Added Uart::availableForWrite()
* Added defines for RAMSTART, RAMSIZE, RAMEND
* Fixed writing LOW to a digital input pin blocking subsequent digitalRead attempts
* Fixed digitalWrite() unnecessarily activating the pull-up resistor
* Wire: Slave writes now use TX buffer
* Added getTimeout() method to Stream.
* Fixed glitch in PWM generation that may happen when calling analogWrite()

cores/arduino/Tone.cpp
/*
  Copyright (c) 2015 Arduino LLC.  All right reserved.

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
  See the GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#include "Arduino.h"
#include "variant.h"

//#define WAIT_TC16_REGS_SYNC(x) while(x->COUNT16.STATUS.bit.PERBUFV);
#define WAIT_TC16_REGS_SYNC(x) while(x->COUNT16.SYNCBUSY.reg & TCC_SYNCBUSY_MASK);


uint32_t toneMaxFrequency = F_CPU / 2;
uint32_t lastOutputPin = 0xFFFFFFFF;

volatile uint32_t *portToggleRegister;
volatile uint32_t *portClearRegister;
volatile uint32_t portBitMask;
volatile int64_t toggleCount;
volatile bool toneIsActive = false;
volatile bool firstTimeRunning = false;

#define TONE_TC         TC4
#define TONE_TC_IRQn    TC4_IRQn
#define TONE_TC_TOP     0xFFFF
#define TONE_TC_CHANNEL 0

void TC4_Handler (void) __attribute__ ((weak, alias("Tone_Handler")));

static inline void resetTC (Tc* TCx)
{
  // Disable TCx
  TCx->COUNT16.CTRLA.reg &= ~TC_CTRLA_ENABLE;
  WAIT_TC16_REGS_SYNC(TCx)

  // Reset TCx
  TCx->COUNT16.CTRLA.reg = TC_CTRLA_SWRST;
  WAIT_TC16_REGS_SYNC(TCx)
  while (TCx->COUNT16.CTRLA.bit.SWRST);
}

void toneAccurateClock (uint32_t accurateSystemCoreClockFrequency)
{
  toneMaxFrequency = accurateSystemCoreClockFrequency / 2;
}

void tone (unsigned char outputPin, unsigned int frequency, unsigned long duration)
{
  
  // Avoid divide by zero error by calling 'noTone' instead
  if (frequency == 0)
  {
    noTone(outputPin);
    return;
  }
  
  // Configure interrupt request
  NVIC_DisableIRQ(TONE_TC_IRQn);
  NVIC_ClearPendingIRQ(TONE_TC_IRQn);
  
  if(!firstTimeRunning)
  {
    firstTimeRunning = true;
    
    NVIC_SetPriority(TONE_TC_IRQn, 0);
      
    // Enable GCLK for TC4  (timer counter input clock)
     GCLK->PCHCTRL[GCM_TC4].reg = ( GCLK_PCHCTRL_CHEN | GCLK_PCHCTRL_GEN_GCLK0 );
     while ( (GCLK->PCHCTRL[GCM_TC4].reg & GCLK_PCHCTRL_CHEN) == 0 ){
	     // wait for sync
	 }
  }
		
  if (toneIsActive && (outputPin != lastOutputPin))
    noTone(lastOutputPin);

  //
  // Calculate best prescaler divider and comparator value for a 16 bit TC peripheral
  //

  uint32_t prescalerConfigBits;
  uint32_t ccValue;

  ccValue = toneMaxFrequency / frequency - 1;
  prescalerConfigBits = TC_CTRLA_PRESCALER_DIV1;
  
  uint8_t i = 0;
  
  while(ccValue > TONE_TC_TOP)
  {
    ccValue = toneMaxFrequency / frequency / (2<<i) - 1;
    i++;
    if(i == 4 || i == 6 || i == 8) //DIV32 DIV128 and DIV512 are not available
     i++;
  }
  
  switch(i-1)
  {
    case 0: prescalerConfigBits = TC_CTRLA_PRESCALER_DIV2; break;
    
    case 1: prescalerConfigBits = TC_CTRLA_PRESCALER_DIV4; break;
    
    case 2: prescalerConfigBits = TC_CTRLA_PRESCALER_DIV8; break;
    
    case 3: prescalerConfigBits = TC_CTRLA_PRESCALER_DIV16; break;
    
    case 5: prescalerConfigBits = TC_CTRLA_PRESCALER_DIV64; break;
      
    case 7: prescalerConfigBits = TC_CTRLA_PRESCALER_DIV256; break;
    
    case 9: prescalerConfigBits = TC_CTRLA_PRESCALER_DIV1024; break;
    
    default: break;
  }

  toggleCount = (duration > 0 ? frequency * duration * 2 / 1000UL : -1LL);

// Disable and reset the TC
   resetTC(TONE_TC);

  uint16_t tmpReg = 0;
  tmpReg |= TC_CTRLA_MODE_COUNT16;  // Set Timer counter Mode to 16 bits
  tmpReg |= prescalerConfigBits; // set prescaler
 // tmpReg |= TC_CTRLA_ONDEMAND;
  TONE_TC->COUNT16.CTRLA.reg |= tmpReg;
  WAIT_TC16_REGS_SYNC(TONE_TC)
  
  TONE_TC->COUNT16.WAVE.reg = TC_WAVE_WAVEGEN_MFRQ; // set counter to frequency match mode
  WAIT_TC16_REGS_SYNC(TONE_TC)

  TONE_TC->COUNT16.CC[TONE_TC_CHANNEL].reg = (uint16_t) ccValue;
  WAIT_TC16_REGS_SYNC(TONE_TC)
  
  // Enable the TONE_TC interrupt request
  TONE_TC->COUNT16.INTENSET.bit.MC0 = 1;
  
  portToggleRegister = &(PORT->Group[g_APinDescription[outputPin].ulPort].OUTTGL.reg);
  portClearRegister = &(PORT->Group[g_APinDescription[outputPin].ulPort].OUTCLR.reg);
  portBitMask = (1ul << g_APinDescription[outputPin].ulPin);

  if (outputPin != lastOutputPin)
  {
    lastOutputPin = outputPin;
    digitalWrite(outputPin, LOW);
    pinMode(outputPin, OUTPUT);
    toneIsActive = true;
  }
  
  // Enable TONE_TC
  TONE_TC->COUNT16.CTRLA.reg |= TC_CTRLA_ENABLE;
  WAIT_TC16_REGS_SYNC(TONE_TC);
  TONE_TC->COUNT16.CTRLBSET.reg |=  TC_CTRLBSET_CMD_RETRIGGER;
  NVIC_EnableIRQ(TONE_TC_IRQn);
}

void noTone (uint8_t outputPin)
{
   /* 'tone' need to run at least once in order to enable GCLK for
   * the timers used for the tone-functionality. If 'noTone' is called
   * without ever calling 'tone' before then 'WAIT_TC16_REGS_SYNC(TCx)'
   * will wait infinitely. The variable 'firstTimeRunning' is set the
   * 1st time 'tone' is set so it can be used to detect wether or not
   * 'tone' has been called before.
   */
  if(firstTimeRunning)
  {
    resetTC(TONE_TC);
    digitalWrite(outputPin, LOW);
    toneIsActive = false;
  }
}

#ifdef __cplusplus
extern "C" {
#endif

void Tone_Handler (void)
{
  if (toggleCount != 0)
  {
    // Toggle the ouput pin
    *portToggleRegister = portBitMask;

    if (toggleCount > 0)
      --toggleCount;

    // Clear the interrupt
    TONE_TC->COUNT16.INTFLAG.bit.MC0 = 1;
  }
  else
  {
    resetTC(TONE_TC);
    *portClearRegister = portBitMask;
    toneIsActive = false;
  }
}

#ifdef __cplusplus
}
#endif
